scaleapi/llm-engine
# Multinode serving

# Pull Request Summary

Multinode serving

API/schema changes:
Bundles/bundle v2:
~Fields in metadata to allow for LWS, specifically we need a different command + some sets of env vars. No code really needed here iiuc since we're just passing through the existing metadata field.~
Use the new fields in the entity for doing multinode

Endpoints:
nodes_per_worker param that controls whether to use a LWS or not
Create LWS endpoint, delete LWS endpoint is allowed
Modifying a LWS endpoint to a non-LWS endpoint or modifying a non-LWS endpoint into a LWS endpoint is not allowed (similar to how we don't allow switching from sync and async) (also changing nodes_per_worker will not be allowed). Enforced by this not being capable through the API.
Get LWS endpoint is allowed
Only available LWS endpoint type will be gpu + streaming, cpu-only LWS doesn't really make sense imo, async/sync could make sense but we don't need it for LLM serving at this point
k8s_erd get_resources + get_all_resources are made LWS compatible
LWS endpoints create a service (separate from the service that the LWS auto-creates) + (if istio is enabled) a ServiceEntry. If istio is enabled, gateway will hackily manually resolve the service's IP and make a request there.

LLM endpoints:
create a multinode bundle if situation calls for it, use endpoint's api to include this new bundle
(TODO reflect in production) build new vllm image to use multinode
Endpoint will do tensorparallel=(# gpus in a single node), pipelineparallel=(# nodes)

Python client: 
Add way to set nodes_per_worker on create + get nodes_per_worker on get/list

## Test Plan and Usage Guide

unit tests
~TODO maybe test for updating existing LWS, _delete_lws doesn't find the lws to delete, _get_all has a LWS returned~
integration tests/e2e tests require LWS, so can't really do this
deploy and manually test some things
ie 
- create llm multinode works, (yes)
- update llm multinode replicas works, (yes)
- get endpoint/llm works, (yes)
- get all endpoint/llm works, (yes)
- delete llm multinode works, (yes)
- request to multinode llm endpoint works (yes)
  - ~currently `curl launch-endpoint-id-end-crtjkptlgk4g03v40o1g.scale-deploy.svc.cluster.local:5005/predict -d '{"prompt": "hihi"}'` from inside another pod in the cluster works, also the pod starts up fine which is cool, requests seem to go through (upd this is because I didn't actually use the second pod's gpus). Port 5000 should also work, probably needs some more formatting.~
  - ~ok I think I have to add another service and name it something different (eww)~
  - ~create other service when LWS is created, name it something different; have sync/streaming inference gateways predict take in a param is_lws to choose between the different service, add a "role: leader" label to the leader template~
  - ~(/workspace/model-engine/model_engine_server/infra/gateways/resources/templates inside a pod)~
  - ~`curl [2602:fb33:0:10:40ff::11]:5000/predict -d '{"args": {"prompt": "hi"}}' -H "Content-Type: application/json"` works from other pod~
  - Multinode starts up (tested with some old llama 2 7b finetune on 2 A10s)
